exchangelib/util.py

```diff
@@ -512,10 +512,15 @@ def post_ratelimited(protocol, session, url, headers, data, allow_redirects=Fals
             # Always create a dummy response for logging purposes, in case we fail in the following
             r = DummyResponse(url=url, headers={}, request_headers=headers)
             try:
-                data = data.decode('utf-8')
-                request = session.prepare_request(Request('POST', url=url, data=data, headers=headers))
-                request.headers['Content-Length'] = str(len(data.encode('utf-8')))
-                r = session.send(request,
+                data = data.encode('utf-8')
+            except UnicodeDecodeError:
+                import chardet
+                encoding_info = chardet.detect(data)
+                data = data.decode('utf-8').encode(encoding_info['encoding'])
```

Comment:
Just to be sure, the order of operations in this case is that we first encode the data based on what chardet detects it to be, and then decode it? Is it possible for chardet to fail to detect?

Reply:
No, we decode it first to turn it into a unicode string, then we encode it with the encoding that chardet detected. It probably is possible for chardet to fail to detect, but we would get an error when sending anyway so :woman_shrugging: 

Encoding without decoding to unicode first fails:
```
>>> x = '•'
>>> x
'\xe2\x80\xa2'
>>> import chardet
>>> chardet.detect(x)
{'confidence': 0.73, 'language': '', 'encoding': 'Windows-1252'}
>>> x.encode('Windows-1252')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/halla/cloud-core/.venv/lib/python2.7/encodings/cp1252.py", line 12, in encode
    return codecs.charmap_encode(input,errors,encoding_table)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 0: ordinal not in range(128)
>>> x.decode('utf-8').encode('Windows-1252')
'\x95'
```

Reply:
Ahhh thanks for the sample! that .73 confidence part is fascinating.